File: app/src/androidTest/java/de/westnordost/streetcomplete/quests/AOsmElementQuestTypeTest.java
package de.westnordost.streetcomplete.quests;

import android.os.Bundle;

import junit.framework.TestCase;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import de.westnordost.streetcomplete.data.osm.OsmElementQuestType;
import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder;
import de.westnordost.streetcomplete.data.osm.changes.StringMapEntryChange;

public abstract class AOsmElementQuestTypeTest extends TestCase
{
	protected Bundle bundle;
	protected Map<String,String> tags;

	public void setUp() throws Exception
	{
		super.setUp();
		bundle = new Bundle();
		tags = new HashMap<>();
	}

	protected final void verify(StringMapEntryChange... expectedChanges)
	{
		StringMapChangesBuilder cb = new StringMapChangesBuilder(tags);
		createQuestType().applyAnswerTo(bundle, cb);
		List<StringMapEntryChange> changes = cb.create().getChanges();
		for(StringMapEntryChange expectedChange : expectedChanges)
		{
			assertTrue(changes.contains(expectedChange));
		}
	}

	protected abstract OsmElementQuestType createQuestType();
}

File: app/src/androidTest/java/de/westnordost/streetcomplete/quests/AOsmElementQuestTypeTest.kt
package de.westnordost.streetcomplete.quests

import android.os.Bundle

import de.westnordost.streetcomplete.data.osm.OsmElementQuestType
import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder
import de.westnordost.streetcomplete.data.osm.changes.StringMapEntryChange
import org.junit.Assert.assertTrue
import org.junit.Before

abstract class AOsmElementQuestTypeTest {

    protected abstract val questType: OsmElementQuestType

    protected var bundle = Bundle()
    protected var tags = mutableMapOf<String, String>()

    @Before fun setUpBundleAndTags() {
        bundle = Bundle()
        tags = mutableMapOf()
    }

    protected fun verify(vararg expectedChanges: StringMapEntryChange) {
        val cb = StringMapChangesBuilder(tags)
        questType.applyAnswerTo(bundle, cb)
        val changes = cb.create().changes
        for (expectedChange in expectedChanges) {
            assertTrue(changes.contains(expectedChange))
        }
    }
}
